tests/Unit/ControlSystem/Actions/Test_LimitTimeStep.cpp

```diff
@@ -0,0 +1,327 @@
+
+  const auto setup_measurements =
+      [&initial_time](
+          const domain::FunctionsOfTime::FunctionOfTime& timescale) {
+        if (timescale.func(0.0)[0][0] ==
+            std::numeric_limits<double>::infinity()) {
+          return control_system::FutureMeasurements(
+              1, std::numeric_limits<double>::infinity());
+        }
+
+        control_system::FutureMeasurements measurements(measurements_per_update,
+                                                        0.0);
+        measurements.update(timescale);
+        while (measurements.next_measurement().value_or(
+                   std::numeric_limits<double>::infinity()) <= initial_time) {
+          measurements.pop_front();
+        }
+        return measurements;
+      };
+
+  auto measurementsA = setup_measurements(*timescales["LabelA"]);
+  auto measurementsBC = setup_measurements(*timescales["LabelBLabelC"]);
+
+  using MockRuntimeSystem = ActionTesting::MockRuntimeSystem<Metavariables>;
+  using component = Component<Metavariables>;
+  MockRuntimeSystem runner{
+      {}, {std::move(timescales), std::move(functions_of_time)}};
+  ActionTesting::emplace_array_component_and_initialize<component>(
+      make_not_null(&runner), ActionTesting::NodeId{0},
+      ActionTesting::LocalCoreId{0}, 0,
+      {std::move(measurementsA), std::move(measurementsBC), std::move(stepper),
+       initial_id, TimeStepId{}, initial_slab.duration(), TimeDelta{},
+       AdaptiveSteppingDiagnostics{}, std::move(history)});
+
+  ActionTesting::set_phase(make_not_null(&runner), Parallel::Phase::Testing);
+  const bool ready =
+      ActionTesting::next_action_if_ready<component>(make_not_null(&runner), 0);
+  if (ready and expected_step_end.has_value()) {
+    const Slab expected_slab(initial_time, *expected_step_end);
+    CHECK(ActionTesting::get_databox_tag<component, Tags::TimeStep>(
+              runner, 0) == expected_slab.duration());
+  } else {
+    CHECK(not ready);
+    CHECK(not expected_step_end.has_value());
+  }
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.ControlSystem.Actions.LimitTimeStep",
+                  "[Unit][ControlSystem]") {
+  register_classes_with_charm<
+      TimeSteppers::AdamsBashforth, TimeSteppers::AdamsMoultonPc,
+      TimeSteppers::Rk3HesthavenSsp,
+      domain::FunctionsOfTime::PiecewisePolynomial<0>>();
+  const double infinity = std::numeric_limits<double>::infinity();
+  const double nan = std::numeric_limits<double>::signaling_NaN();
+  const double arbitrary = -1234.0;
+
+  // clang-format off
+  test("No control systems", 1.0, 5.0, 5.0,
+       {{0.0, infinity}, {infinity, nan}},
+       {{0.0, arbitrary}, {infinity, nan}},
+       {{0.0, infinity}, {infinity, nan}},
+       {{0.0, arbitrary}, {infinity, nan}},
+       {{0.0, arbitrary}, {infinity, nan}});
+  test("Step much shorter than limit", 1.0, 5.0, 5.0,
+       {{0.0, 10.0}, {50.0, nan}},
+       {{0.0, arbitrary}, {50.0, nan}},  // goal range [20, 50]
```

Comment:
What is this goal range? The 50 seems somewhat self explanatory, but the 20 doesn't. Could you add a comment before this about what the goal range is supposed to represent?